### ISSUE TITLE
Very low mIOU

### ISSUE THREAD

**Issue by arslan-chaudhry** [DrSleep/tensorflow-deeplab-resnet#18]
Training from scratch using the initialization model - deeplab_resnet_init.ckpt (provided in the instructions) - and default parameters (learning rate, number of iterations etc) results in a network which gives very low mIOU (**53.2%**) on the validation set. Whereas, pre-trained model (deeplab_resnet.ckpt, provided in the instructions) gives 78.9% mIOU. I am just wondering, if you used the same parameters (20k iterations, 1e-4 learning rate etc...) for training the model which gives 78.9% on validation set. 

P.S: I didn't run fine_tune.py. I just ran train.py, followed by evaluate.py.

**Comment by DrSleep** [comment]
The pre-trained model is the one provided by the authors, i.e. pre-trained by them. 
The provided training parameters are not optimal. I have used them for fine-tuning only, which leads to around 70% mIoU after training only on Pascal VOC (the original paper also makes use of the COCO dataset, as well).

**Comment by arslan-chaudhry** [comment]
Fine-tuning the trained model (trained from scratch using train.py, with deeplab_resnet_init.ckpt as initialization model) with the default parameters provided in fine_tune.py results in **59.7%** mIoU. How did you get close to 70%?

**Comment by DrSleep** [comment]
I have only used the fine-tuning script to get close to 70% (in another issue, the result 65.3% has been mentioned - #6).

**Comment by arslan-chaudhry** [comment]
Which model did you fine-tune? The pre-trained one from the authors?

**Comment by DrSleep** [comment]
The init model.

**Comment by arslan-chaudhry** [comment]
So training with momentum with "poly" learning rate policy as described in the paper, adding weight decay, and running the training for 40k iterations has increased the mIoU to 69%. Due to my GPU memory bottleneck I can't increase the batch_size beyond 4. Maybe that accounts for the still missing 3-4% accuracy.  
Previously, I tried with default training parameters in this repo, and even after 50k iterations the accuracy on validation set was only 53%.

**Comment by mgarbade** [comment]
@arslan-chaudhry could you share the info how you applied "poly" and "weight_decay" in tensorflow?
My performance is still at 66 % (finetuning the entire network for 20k iterations, Adam optimizer with a learning rate of 1e-4 for the last layer and 1e-6 for the rest of the network, only convolutions and atrous convolutions are trained, all other parameters are kept fix)

**Comment by DrSleep** [comment]
I have re-written the training script following the original choices for optimisation hyperparameters (SGD with momentum, weight decay, learning rate with polynomial decay, different learning rates for different layers), and ignoring the void label during the training. 
At the moment, the code resides here: https://github.com/DrSleep/tensorflow-deeplab-resnet/tree/train-orig; I will push it to the master branch once I clean it and finish the training procedure.
So far, after 5000 iterations (with batch_size=10) mean IoU is 72.2 %.

**Comment by gaopeng-eugene** [comment]
72.2% is the result trained from deeplab_resnet.ckpt or deeplab_resnet_init.ckpt ?

**Comment by gaopeng-eugene** [comment]
I think the batch size for the original model is 1 which means batch size will not affect the mAP too much.

**Comment by DrSleep** [comment]
deeplab_resnet_init.ckpt; in the original model the effective batch size is 10 (batch_size * iter_size).

**Comment by arslan-chaudhry** [comment]
Did you set the --is-training flag, in other words, are you changing the batch norm statistics given your batch size is now 10?

**Comment by arslan-chaudhry** [comment]
Previously when I was trying to change the batch size from 4 to 10, I was running into out of memory error, but now with the new training script batch size of 10 is not producing out of memory error any more. Can you please comment what could be the reason for such a behaviour? 
Also, do you think it is worthwhile to implement the gradient accumulation strategy of the original paper (i.e) accumulating the gradients for 10 iterations and then applying it at once. I know with no corresponding implementation of "iter_size" in tensorflow the code will be a bit messy but then GPUs with low memory can make use of this code.

Also, I see that you are not training the batch norm parameters any more? Any specific reason for that?

**Comment by arslan-chaudhry** [comment]
I can't seem to reproduce your results. I tried the new training script. After training the mIoU on validation set is 48.4% (surprisingly lower than the previous model). I am using initialization model that I converted from the **init.caffemodel** provided by the authors. One thing different from the repo is that I am doing random horizontal flipping of the images and labels as well, along with scaling. I didn't pass "--is-training" flag to the training script, I used:
$> python train.py --random-scale

**Comment by arslan-chaudhry** [comment]
Using **deeplab_resnet_init.ckpt** and without horizontal flipping the mIoU is **70.9%** after 20k iterations. If I had to guess, there is something wrong with the init model that I converted manually from init.caffemodel, but I am not sure as I followed the exact instructions given in the README.

Also, still the mIoU is quite low. How come you end up with 72.2% after 5k iterations? What is your final accuracy after 20k iteration?

**Comment by DrSleep** [comment]
The final accuracy is 72.7. 
I haven't set the is_training flag.

**Comment by DrSleep** [comment]
I am unable to reproduce your problems with init.caffemodel. I have run the same pipeline once again, and got similar results as before.
Could you tell me please whether you use PyCaffe when converting? And what is the size of the produced file?

**Comment by arslan-chaudhry** [comment]
I am using **convert.py** script provided in the repo to extract weights from init.caffemodel, followed by the use of **npy2ckpt.py** to get the weights in TF format. The size of the file is 169Mb. 

But, as I said in one of my comments earlier, if I use the **deeplab_resnet_init.ckpt** provided by you, even then the final accuracy is 70.9%, whereas you are getting 72.7?

**Comment by DrSleep** [comment]
Well, I cannot reproduce any of your problems.  

And even if you use the convert.py script but without PyCaffe, such a regime is not tested and you would see the following error: https://github.com/ethereon/caffe-tensorflow/blob/0c5bf53817f89c7e4581ad6b738a4bb19eea96b3/kaffe/caffe/resolver.py#L41

**Comment by arslan-chaudhry** [comment]
I am using PyCaffe. Anyhow, let's ignore the manual conversion of init.caffemodel for now. Right now, I am using the same **deeplab_resnet_init.ckpt** model for initialization. With your latest commit, I was able to bump up the accuracy to **71.6%** which is still below than what are you are seeing: **72.7%**. Can you please share your trained model somewhere?
Also, both of these accuracies are still way below than original model. I know, the implementation is still missing multi-scaling and CRF post-processing, but I don't think these should create a delta of **6.2%**. What are your thoughts?

**Comment by arslan-chaudhry** [comment]
Also, any specific reason you are not training batchnorm parameters (beta, gamma)?

**Comment by mgarbade** [comment]
This is the exact same training configuration as in the script provided by the deeplab people as I described [here](https://github.com/DrSleep/tensorflow-deeplab-resnet/issues/12#issuecomment-276707835). The reason might be that beta, gamma and all batch norm parameters are hard not trainable for small batch sizes.

**Comment by DrSleep** [comment]
I believe that the difference is due to the fact that the original model has 3 paths (for an image in original resolution, 0.5 downscaled and 0.75 downscaled) and 4 losses (original resolution, 0.5 resolution, 0.75 resolution and the fusion of all outputs).
The batch norm parameters are preserved due to a relatively small batch size. 

I will upload the model later.

**Comment by arslan-chaudhry** [comment]
It might be worth considering randomizing the crop of images instead of always doing a centre-crop if the size exceeds 321. I believe, this is what the original paper has done.

**Comment by DrSleep** [comment]
Here is the link to the model after 20000 iterations (I have trained it once again - mIoU is 72.6): https://drive.google.com/open?id=0B_rootXHuswsVHBfN2lYUmFYQ0E

**Comment by amiltonwong** [comment]
Hi, @DrSleep and @arslan-chaudhry ,

My concern is : the weight file : [deeplab_resnet.ckpt](https://drive.google.com/drive/folders/0B_rootXHuswsZ0E4Mjh1ZU5xZVU) is already a final one, it has been trained well. And we could simply use it for evaluation (val dataset) and get high score (mIoU: 0.869)

Here is the evaluation output:
```
(python2.7_tf0.12.1_gpu) root@milton-All-Series:/data/code/tensorflow-deeplab-resnet# python evaluate.py
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcurand.so locally
I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 0 with properties: 
name: GeForce GTX 980 Ti
major: 5 minor: 2 memoryClockRate (GHz) 1.2405
pciBusID 0000:01:00.0
Total memory: 5.93GiB
Free memory: 5.36GiB
W tensorflow/stream_executor/cuda/cuda_driver.cc:590] creating context when one is currently active; existing: 0x44ca080
I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 1 with properties: 
name: GeForce GTX 980 Ti
major: 5 minor: 2 memoryClockRate (GHz) 1.2405
pciBusID 0000:08:00.0
Total memory: 5.93GiB
Free memory: 5.83GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0 1 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 1:   Y Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 980 Ti, pci bus id: 0000:01:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:1) -> (device: 1, name: GeForce GTX 980 Ti, pci bus id: 0000:08:00.0)
Restored model parameters from ./deeplab_resnet.ckpt
step 0
step 100
step 200
step 300
step 400
step 500
step 600
step 700
step 800
step 900
step 1000
step 1100
step 1200
step 1300
step 1400
Mean IoU: 0.869

```

Please correct me if I'm wrong ~

**Comment by arslan-chaudhry** [comment]
@DrSleep your model (https://drive.google.com/drive/folders/0B_rootXHuswsVHBfN2lYUmFYQ0E) is giving me **71.5%**.
 
mac@mac-linux:~/phd/projects/tensorflow-deeplab-resnet$ python evaluate.py 
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcurand.so locally
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 0 with properties: 
name: TITAN X (Pascal)
major: 6 minor: 1 memoryClockRate (GHz) 1.531
pciBusID 0000:01:00.0
Total memory: 11.89GiB
Free memory: 10.70GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: TITAN X (Pascal), pci bus id: 0000:01:00.0)
Restored model parameters from /home/mac/Downloads/drsleep_trained/model.ckpt-20000
step 0
step 100
step 200
step 300
step 400
step 500
step 600
step 700
step 800
step 900
step 1000
step 1100
step 1200
step 1300
step 1400
Mean IoU: 0.715

**Comment by arslan-chaudhry** [comment]
@amiltonwong it looks dubious. On PASCAL VOC leaderboard (http://host.robots.ox.ac.uk:8080/leaderboard/displaylb.php?challengeid=11&compid=6) PSPNet has 85.4% mIoU (the highest). The reported accuracy of DeeplabV2 after CRF post-processing is 79.7%. It can't be 86.9.

**Comment by amiltonwong** [comment]
@arslan-chaudhry , the 86.9 score is for validation data. The leaderboard only shows score for test data. 
I'm also a bit confused when the 86.9 score comes out :P

**Comment by arslan-chaudhry** [comment]
@amiltonwong the paper reports the accuracy on validation set in the range of 77-78%. 86.9 still doesn't make any sense.

**Comment by DrSleep** [comment]
I think the reason we are reporting different numbers is that we are using different datasets, in particular, different SegmentationClassAug: I am not sure about the link provided by @amiltonwong (although it has also given me 86.9%) - https://www.dropbox.com/s/oeu149j8qtbs1x0/SegmentationClassAug.zip?dl=0, but I have been following the official DeepLab FAQ and using this dataset -http://home.bharathh.info/pubs/codes/SBD/download.html - converted to raw png files.

I will try to upload the dataset that I am using so we can all report results on the same data.
In the meanwhile, if someone can give a try to the multi-scale version - https://github.com/DrSleep/tensorflow-deeplab-resnet/tree/msc - and train it until convergence from the init model (probably, increasing the number of steps as the batch size is smaller), that would be really helpful.

**Comment by arslan-chaudhry** [comment]
@DrSleep I am using the dataset from the same link as you provided. I think it is a good idea to have a dataset uploaded somewhere. 
I'll try out the msc branch.

**Comment by amiltonwong** [comment]
@DrSleep @arslan-chaudhry ,
Yes, I use this [link](https://www.dropbox.com/s/oeu149j8qtbs1x0/SegmentationClassAug.zip?dl=0) provided by this [thread](https://github.com/TheLegendAli/DeepLab-Context/issues/10#issuecomment-255064521)

**Comment by DrSleep** [comment]
I have sent the results on the validation set to the official evaluation server of PASCAL VOC: http://host.robots.ox.ac.uk/anonymous/FIQPRH.html
The score is 86.9%, which means that we should use the dataset provided by @amiltonwong (thanks for the link!). 

I have also checked why there are discrepancies between our results: turned out, that in my case the void '255' label is not replaced by '0' in some images, which reduces the overall score as in the evaluation script the void label is ignored while the '0' label is penalised if predicted wrong: https://gist.github.com/DrSleep/cd6d1248958b92d1a46a7a71b61a1568.
@arslan-chaudhry, I guess your dataset has similar problems as mine. Please try to use the new one.

**Comment by gaopeng-eugene** [comment]
why the mAP is 86.9%? it's better than the state of the art !!

**Comment by DrSleep** [comment]
Well, it is only on the validation set (on the test set the results are more realistic - around 76% mIoU).
I think the results on the validation set mentioned in the DeepLab paper are without ignoring the void label.

**Comment by arslan-chaudhry** [comment]
@DrSleep so training with **msc** branch resulted in the accuracy of **74.7%**. I increased the number of iterations to **40k**. This is using the dataset pointed by @amiltonwong where pre-trained model is achieving 86.9%. Also, on the same dataset if I use the model trained without msc, I get **74.4%**. So currently msc is not doing a lot.
I think this has something to do with the batch size. As I alluded to earlier, I think, we should implement the gradient accumulation strategy for multiple iterations (10 in our case), as implemented in the original paper. This will allow us to experiment with bigger batch sizes without running into memory bottleneck.

**Comment by arslan-chaudhry** [comment]
@DrSleep have you tried experimenting with **msc** branch? 
Right now, I am trying to accumulate the gradients over 10 iterations with a batch_size of 1, and then making a single gradient update. For some reason, my gradients are blowing up after a first few iterations. I am trying to debug that. I still think following this strategy will make things more scalable. What are your thoughts?

**Comment by mgarbade** [comment]
Did you average the gradients or just adding them up? By going through the Caffe implementation today I saw that they are added with a weight of 1 / iter_size as mentioned [here](https://github.com/BVLC/caffe/blob/536851c00e44d545649648b372503f52a89c0499/src/caffe/solvers/sgd_solver.cpp#L123)

**Comment by arslan-chaudhry** [comment]
@mgarbade yes I am.

**Comment by DrSleep** [comment]
@arslan-chaudhry, I have run the training only for 3000 iterations, so I haven't fully trained it. 

Gradients accumulation should be useful, I agree; when I have time, will try to implement it. 
I remember I was using similar to [this approach](https://github.com/tensorflow/tensorflow/issues/3994) on a different task, and averaging gradients prevented the training process from diverging.

**Comment by arslan-chaudhry** [comment]
@DrSleep I am using the similar approach. I was able to spot the bug which was causing the divergence. With gradient accumulation after 11k iterations I am getting **75.5%** accuracy. This is after fixing the image crop issue reported by @mgarbade.

**Comment by arslan-chaudhry** [comment]
@DrSleep after training for 20k iteration using gradient accumulation, I am getting **76.78%** on **test data**: http://host.robots.ox.ac.uk:8080/anonymous/JL6OTV.html. 
Whereas, if I use the **pre-trained** model (deeplab_resnet.ckpt) the results on the test set are **76.65%**: http://host.robots.ox.ac.uk:8080/anonymous/C9TN23.html

**Comment by gaopeng-eugene** [comment]
@arslan-chaudhry can you summarize how you train the model from scratch?

**Comment by arslan-chaudhry** [comment]
@gaopeng-eugene in addition to what has already been implemented for multi-scale input in this repository, I used random cropping instead of centre crop.  I used **deeplab_resnet_init.ckpt** as the initialization model. Additionally, I used gradient accumulation strategy, where I stored gradients for 10 iterations and then made a single gradient update effectively making the batch_size 10:
```
# Define a variable to accumulate gradients
accum_grads = [tf.Variable(tf.zeros_like(v.initialized_value()),
                           trainable=False) for v in conv_trainable + fc_w_trainable + fc_b_trainable]

# Define an operation to clear the accumulated gradients for next batch
zero_op = [v.assign(tf.zeros_like(v)) for v in accum_grads]

# Compute gradients
grads = tf.gradients(reduced_loss, conv_trainable + fc_w_trainable + fc_b_trainable)

accum_grads_op = [accum_grads[i].assign_add(grad) for i, grad in
                   enumerate(grads)]

# Normalize the gradients before applying
grads_conv = np.asarray(accum_grads[:len(conv_trainable)]) / GRAD_UPDATE_EVERY
grads_fc_w = np.asarray(accum_grads[len(conv_trainable) : (len(conv_trainable) + len(fc_w_trainable))]) / GRAD_UPDATE_EVERY
grads_fc_b = np.asarray(accum_grads[(len(conv_trainable) + len(fc_w_trainable)):]) / GRAD_UPDATE_EVERY

# Apply the gradients
train_op_conv = opt_conv.apply_gradients(zip(grads_conv, conv_trainable))
train_op_fc_w = opt_fc_w.apply_gradients(zip(grads_fc_w, fc_w_trainable))
train_op_fc_b = opt_fc_b.apply_gradients(zip(grads_fc_b, fc_b_trainable))

train_op = tf.group(train_op_conv, train_op_fc_w, train_op_fc_b)
```
The corresponding iterations over training steps are:
```
# Iterate over training steps.
for step in range(args.num_steps):
    start_time = time.time()
    feed_dict = { step_ph : step }
    loss_value = 0

    # Clear the accumulated gradients
    sess.run(zero_op, feed_dict=feed_dict)

    # Accumulate gradients
    for i in range(GRAD_UPDATE_EVERY):
        _, l_val = sess.run([accum_grads_op, reduced_loss], feed_dict=feed_dict)
        loss_value += l_val

    # Normalize the loss
    loss_value /= GRAD_UPDATE_EVERY

    # Apply gradients
    if step % args.save_pred_every == 0:
        images, labels, summary, _ = sess.run([image_batch, label_batch, total_summary, train_op], feed_dict=feed_dict)
        summary_writer.add_summary(summary, step)
        save(saver, sess, args.snapshot_dir, step)
    else:
        sess.run(train_op, feed_dict=feed_dict)

    duration = time.time() - start_time
    print('step {:d} \t loss = {:.3f}, ({:.3f} sec/step)'.format(step,
                                                                 loss_value, duration))
```

**Comment by gaopeng-eugene** [comment]
Can you tell me the best val result and loss for the last epoch.

**Comment by arslan-chaudhry** [comment]
On val I get **76.5%**. I don't have the training log, but IIRC, the loss in the final steps is somewhere near 1.2-1.5.

**Comment by arslan-chaudhry** [comment]
As for 86.9% on validation set from pre-trained model, my hunch is that the model is trained on **train+val** dataset.

**Comment by gaopeng-eugene** [comment]
The gap between testing and validation is so small. Have you ever tried the deeplab vgg model? The loss is about 0.1. Why there is such a big difference between the vgg deeplab and resnet deeplab?

**Comment by gaopeng-eugene** [comment]
Another issue. In my experiment the pretained model only achieve a 0.801 on validation set.
CUDA_VISIBLE_DEVIC
ES=6 python evaluate.py --restore-from ../tensorflow-deeplab-resnet3/train2model/model.ckpt 
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcurand.so locally
I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 0 with properties: 
name: Tesla P100-SXM2-16GB
major: 6 minor: 0 memoryClockRate (GHz) 1.4805
pciBusID 0000:89:00.0
Total memory: 15.93GiB
Free memory: 15.64GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla P100-SXM2-16GB, pci bus id: 0000:89:00.0)
Restored model parameters from ../tensorflow-deeplab-resnet3/train2model/model.ckpt
step 0
step 100
step 200
step 300
step 400
step 500
step 600
step 700
step 800
step 900
step 1000
step 1100
step 1200
step 1300
step 1400
Mean IoU: 0.801

**Comment by gaopeng-eugene** [comment]
train2-iter-20000.caffemodel -> train2model/model.ckpt

**Comment by arslan-chaudhry** [comment]
Are you using the dataset pointed out by @amiltonwong above? See that in your dataset you are not setting the ignore_labels (255) to 0.

**Comment by gaopeng-eugene** [comment]
I am using the validation set in this repo. By the way, could u share the random cropping code and the performance difference between random crop and central crop?

**Comment by arslan-chaudhry** [comment]
See that last comment on [12](https://github.com/DrSleep/tensorflow-deeplab-resnet/issues/12) from @mgarbade. I have not tried to train with centre crop and gradient accumulation, so I don't have that datapoint.

**Comment by DrSleep** [comment]
@arslan-chaudhry, nice results with gradients accumulation and random crop! 
Would be helpful to have it as a pull request as well :)

**Comment by arslan-chaudhry** [comment]
@DrSleep sure thing. Do you plan to merge the **msc** branch to master?

**Comment by arslan-chaudhry** [comment]
https://github.com/DrSleep/tensorflow-deeplab-resnet/pull/30
Review and merge at your convenience.

**Comment by DrSleep** [comment]
@arslan-chaudhry, thank you!

**Comment by arslan-chaudhry** [comment]
@DrSleep I have sent a PR https://github.com/DrSleep/tensorflow-deeplab-resnet/pull/34 for adding **evaluate_msc.py** script in the master branch. 
I also think it is worthwhile to add CRF post processing in the master branch, so that we have the whole pipeline ready on tensorflow.

**Comment by DrSleep** [comment]
@arslan-chaudhry, thanks, I will look at it later!
As for CRF, I have used it with the same parameters as in the original model, but it has given me lower scores. Besides that, the evaluation with CRF takes significantly more time (IIRC ~x10-15). That's why I don't think it is yet ready to be merged

**Comment by gaopeng-eugene** [comment]
In my experiment, fixing the variance and mean of batch norm layer can achieve better results. How about your results?

**Comment by gaopeng-eugene** [comment]
@arslan-chaudhry Thank you so much for the sharing of multi scale training. However, in my experiment, the performance is much worse than the single training.

Here is the parameter:

BATCH_SIZE = 2
GRAD_UPDATE_EVERY = 5

After 15000 iterations, the evaluate performance is 0.498.

Best Wishes
